src/aapns/pool.py

```diff
@@ -0,0 +1,218 @@
+from __future__ import annotations
+
+import asyncio
+import ssl
+from asyncio import (
+    CancelledError,
+    Event,
+    TimeoutError,
+    create_task,
+    gather,
+    sleep,
+    wait_for,
+)
+from contextlib import contextmanager, suppress
+from dataclasses import dataclass, field
+from itertools import count
+from logging import getLogger
+from random import shuffle
+from time import time
+from typing import Optional, Set
+
+from .connection import Connection, Request, Response, create_ssl_context
+from .errors import Blocked, Closed, FormatError, ResponseTooLarge, Timeout
+
+logger = getLogger(__package__)
+
+
+@dataclass(eq=False)
+class Pool:
+    """Simple fixed-size connection pool with automatic replacement"""
+
+    origin: str
+    size: int
+    ssl_context: ssl.SSLContext
+    active: Set[Connection]
+    dying: Set[Connection] = field(default_factory=set)
+    closing: bool = False
+    closed: bool = False
+    errors: int = 0
+    retrying: int = 0
+    completed: int = 0
+    outcome: Optional[str] = None
+    maintenance: asyncio.Task = field(init=False)
+    maintenance_needed: asyncio.Event = field(default_factory=asyncio.Event)
+
+    def __repr__(self):
+        bits = [
+            self.state,
+            self.origin,
+            f"alive:{len(self.active)}",
+            f"dying:{len(self.dying)}",
+        ]
+        if self.state != "closed":
+            all = self.active | self.dying
+            bits.append(f"buffered:{sum(c.buffered for c in all)}")
+            bits.append(f"inflight:{sum(c.inflight for c in all)}")
+        bits.append(f"retrying:{self.retrying}")
+        bits.append(f"completed:{self.completed}")
+        bits.append(f"errors:{self.errors}")
+        return "<Pool %s>" % " ".join(bits)
+
+    @classmethod
+    async def create(cls, origin: str, size=2, ssl=None):
+        if size < 1:
+            raise ValueError("Connection pool size must be strictly positive")
+        ssl_context = ssl or create_ssl_context()
+        connections = set(
+            await gather(
+                *(Connection.create(origin, ssl=ssl_context) for i in range(size))
+            )
+        )
+        # FIXME run the hook / ensure no connection is dead
+        return cls(origin, size, ssl_context, connections)
+
+    def __post_init__(self):
+        self.maintenance = create_task(self.maintain(), name="maintenance")
+
+    @property
+    def state(self):
+        if not self.closing:
+            return "active"
+        elif not self.closed:
+            return "closing"
+        else:
+            return "closed"
+
+    def resize(self, size):
+        assert size > 0
+        self.size = size
+        self.maintenance_needed.set()
+
+    def cert_hook(self, c):
+        if not self.outcome and c.outcome == "BadCertificateEnvironment":
+            self.closing = True
+            self.outcome = c.outcome
+
+    async def maintain(self):
```

Comment:
please use `connection` (or if that's too long, `conn`) rather than `c` in this method.